applications/mne_analyze/plugins/coregistration/coregistration.cpp

```diff
@@ -0,0 +1,756 @@
+
+void CoRegistration::unload()
+{
+}
+
+//=============================================================================================================
+
+QString CoRegistration::getName() const
+{
+    return "Co-Registration";
+}
+
+//=============================================================================================================
+
+QMenu *CoRegistration::getMenu()
+{
+    return Q_NULLPTR;
+}
+
+//=============================================================================================================
+
+QDockWidget *CoRegistration::getControl()
+{
+    // Coregistration Settings
+    m_pCoregSettingsView = new CoregSettingsView(QString("MNEANALYZE/%1").arg(this->getName()));
+
+    QDockWidget* pControlDock = new QDockWidget(getName());
+    pControlDock->setAllowedAreas(Qt::RightDockWidgetArea | Qt::LeftDockWidgetArea);
+    pControlDock->setWidget(m_pCoregSettingsView);
+    pControlDock->setObjectName(getName());
+
+    // connect buttons
+    connect(m_pCoregSettingsView, &CoregSettingsView::changeSelectedBem,
+            this, &CoRegistration::onChangeSelectedBem, Qt::UniqueConnection);
+    connect(m_pCoregSettingsView, &CoregSettingsView::digFileChanged,
+            this, &CoRegistration::onDigitizersChanged);
+    connect(m_pCoregSettingsView, &CoregSettingsView::fidFileChanged,
+            this, &CoRegistration::onFiducialsChanged);
+    connect(m_pCoregSettingsView, &CoregSettingsView::fitFiducials,
+            this, &CoRegistration::onFitFiducials);
+    connect(m_pCoregSettingsView, &CoregSettingsView::fitICP,
+            this, &CoRegistration::onFitICP);
+    connect(m_pCoregSettingsView, &CoregSettingsView::fidStoreFileChanged,
+            this, &CoRegistration::onStoreFiducials);
+    connect(m_pCoregSettingsView, &CoregSettingsView::transParamChanged,
+            this, &CoRegistration::onUpdateTrans);
+    connect(m_pCoregSettingsView, &CoregSettingsView::loadTrans,
+            this, &CoRegistration::onLoadTrans);
+    connect(m_pCoregSettingsView, &CoregSettingsView::storeTrans,
+            this, &CoRegistration::onStoreTrans);
+
+    // connect fiducial selection
+    connect(m_pCoregSettingsView, &CoregSettingsView::pickFiducials,
+            this, &CoRegistration::onPickFiducials);
+    connect(m_pCoregSettingsView, &CoregSettingsView::fiducialChanged,
+            this, &CoRegistration::onPickedFiducialChanged);
+
+    // Connect events for new and deleted model
+    connect(m_pAnalyzeData.data(), &AnalyzeData::newModelAvailable,
+            this, &CoRegistration::updateBemList, Qt::UniqueConnection);
+    connect(m_pAnalyzeData.data(), &AnalyzeData::modelIsEmpty,
+            this, &CoRegistration::deleteModels, Qt::UniqueConnection);
+
+    // automaticly load Bem if available
+    onChangeSelectedBem(m_pCoregSettingsView->getCurrentSelectedBem());
+
+    // Thread handling
+    connect(&m_FutureWatcher, &QFutureWatcher<QMap<double,QList<int>>>::finished,
```

Comment:
@gabrielbmotta Can you check whether this PR uses the multithreading correctly?

Reply:
I'm not sure about the thread safety of the 'get' functions on lines 540-549 reading directly from the GUI widgets on a separate thread, but at the same time I don't think its possible to change them fast enough to be a problem. We could have either the coregsettingsview or the coregistration store those values as members to be updated by the gui with signals, and have the updating and the reading mutexed.

Reply:
I'm now saving the values as member in coregistratio and use a mutex to set the values befor starting icp.